cmd/daemon/middleware.go

```diff
@@ -34,3 +37,33 @@ func NewNegroniLoggerMiddleware(l *logrusx.Logger, name string) *reqlog.Middlewa
 	}
 	return n
 }
+
+func NewStripPrefixMiddleware(prefixLoader func(r *http.Request) string) negroni.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+		prefix := strings.TrimRight(prefixLoader(r), "/")
+		if len(prefix) == 0 {
+			next.ServeHTTP(w, r)
+			return
+		}
+		http.StripPrefix(prefix, next).ServeHTTP(w, r)
+	}
+}
+
+func extractPrefixFromBaseURL(u *url.URL) string {
+	if u == nil {
+		return ""
+	}
+	return strings.TrimSuffix(u.Path, "/")
+}
+
+func publicURLPrefixExtractor(provider config.Provider) func(r *http.Request) string {
+	return func(r *http.Request) string {
+		return extractPrefixFromBaseURL(provider.Config(r.Context()).SelfPublicURL(r))
```

Comment:
Wouldn't it make more sense to do this in `SelfPublicURL` / `SelfAdminURL` so that the result is the same across the code base?